go/tasks/pluginmachinery/flytek8s/pod_helper.go

```diff
@@ -145,25 +146,58 @@ func BuildPodWithSpec(podTemplate *v1.PodTemplate, podSpec *v1.PodSpec, primaryC
 	}
 
 	if podTemplate != nil {
+		// merge template PodSpec
 		basePodSpec := podTemplate.Template.Spec.DeepCopy()
 		err := mergo.Merge(basePodSpec, podSpec, mergo.WithOverride, mergo.WithAppendSlice)
 		if err != nil {
 			return nil, err
 		}
 
+		// merge template Containers
 		var mergedContainers []v1.Container
+		var defaultContainerTemplate, primaryContainerTemplate *v1.Container
+		for i := 0; i < len(podTemplate.Template.Spec.Containers); i++ {
+			if podTemplate.Template.Spec.Containers[i].Name == defaultContainerTemplateName {
+				defaultContainerTemplate = &podTemplate.Template.Spec.Containers[i]
+			} else if podTemplate.Template.Spec.Containers[i].Name == primaryContainerTemplateName {
+				primaryContainerTemplate = &podTemplate.Template.Spec.Containers[i]
+			}
+		}
+
 		for _, container := range podSpec.Containers {
-			for _, templateContainer := range basePodSpec.Containers {
-				if ((container.Name == primaryContainerName) && templateContainer.Name == DefaultContainerName) || (container.Name == templateContainer.Name) {
-					newContainer := templateContainer.DeepCopy()
-					err := mergo.Merge(newContainer, container)
+			// if applicable start with defaultContainerTemplate
+			var mergedContainer *v1.Container
+			if defaultContainerTemplate != nil {
+				mergedContainer = defaultContainerTemplate.DeepCopy()
+			}
+
+			// if applicable merge with primaryContainerTemplate
+			if container.Name == primaryContainerName && primaryContainerTemplate != nil {
+				if mergedContainer == nil {
+					mergedContainer = primaryContainerTemplate.DeepCopy()
+				} else {
+					err := mergo.Merge(mergedContainer, primaryContainerTemplate, mergo.WithOverride, mergo.WithAppendSlice)
```

Comment:
I don't remember the specific example anymore, but I remember I removed the `WithOverride` option because I ran into an issue where the container name got overridden at times I didn't want it to be. Maybe this won't be a problem now since the logic is structured a little differently though!